ipa-core/src/helpers/gateway/receive.rs

```diff
@@ -1,36 +1,61 @@
-use std::marker::PhantomData;
+use std::{
+    marker::PhantomData,
+    pin::Pin,
+    task::{Context, Poll},
+};
 
 use bytes::Bytes;
 use dashmap::{mapref::entry::Entry, DashMap};
+use futures::Stream;
+use pin_project::pin_project;
 
 use crate::{
     error::BoxError,
     helpers::{
-        buffers::UnorderedReceiver, gateway::transport::RoleResolvingTransport, Error,
-        HelperChannelId, LogErrors, Message, Role, Transport,
+        buffers::UnorderedReceiver, gateway::transport::RoleResolvingTransport,
+        transport::SingleRecordStream, ChannelId, Error, HelperChannelId, LogErrors, Message,
+        MpcMessage, Role, ShardChannelId, ShardTransportImpl, Transport, TransportIdentity,
     },
     protocol::RecordId,
+    sync::{Arc, Mutex},
 };
 
-/// Receiving end of the gateway channel.
-pub struct ReceivingEnd<M: Message> {
+/// Receiving end of the MPC gateway channel.
+/// I tried to make it generic and work for both MPC and Shard connectors, but ran into
+/// "implementation of `S` is not general enough" issue on the client side (reveal). It may be another
+/// occurrence of [`gat`] issue
+///
+/// [`gat`]: https://github.com/rust-lang/rust/issues/100013
+pub struct MpcReceivingEnd<M> {
     channel_id: HelperChannelId,
     unordered_rx: UR,
     _phantom: PhantomData<M>,
 }
 
+#[pin_project]
+pub struct ShardReceivingEnd<M: Message> {
+    pub(super) channel_id: ShardChannelId,
+    #[pin]
+    pub(super) rx: SingleRecordStream<M, ShardReceiveStream>,
+}
+
 /// Receiving channels, indexed by (role, step).
-#[derive(Default)]
-pub(super) struct GatewayReceivers {
-    pub(super) inner: DashMap<HelperChannelId, UR>,
+pub(super) struct GatewayReceivers<I, S> {
+    pub(super) inner: DashMap<ChannelId<I>, S>,
 }
 
-pub(super) type UR = UnorderedReceiver<
+pub type UR = UnorderedReceiver<
     LogErrors<<RoleResolvingTransport as Transport>::RecordsStream, Bytes, BoxError>,
     Vec<u8>,
 >;
 
-impl<M: Message> ReceivingEnd<M> {
+/// Stream of records received from a peer shard.
+#[derive(Clone)]
+pub struct ShardReceiveStream(
+    pub(super) Arc<Mutex<<ShardTransportImpl as Transport>::RecordsStream>>,
```

Comment:
Does `Arc<Mutex>` make the most sense here, or does something more like a `RefCell` make more sense? I'm thinking that multiple concurrent readers aren't going to do something sensible, so maybe we should try to prevent that.

Reply:
I think stall detection and MPC receiver working in parallel may be a valid use case here. It does not seem that `RefCell` can work here even if we don't have concurrent readers - it requires borrowing from behind a Dashmap mutex which is a bad idea and hopefully not possible

Reply:
Having the `Arc<Mutex>` within `ShardReceiveStream` means that every poll operation on a shard receive stream (even when built without stall detection) will pay a cost to lock the mutex.

I did spend a little time trying to figure out a concrete solution, but I wasn't successful. The ShardReceiveStream can hold a reference to a record counter in the gateway state (rather than the gateway state holding a reference to the receive stream), but holding dashmap entry references directly doesn't work. Maybe both the ShardReceiveStream and the gateway state could hold an `Arc<AtomicU32>`?

It may not be worth the effort to solve this right now. But I think it would at least be worth a comment that there is a potential optimization here (definitely in the case where stall detection is disabled, and possibly even in the case where it is enabled).

Reply:
Yea that's not great, although this is not the only place where we lock the mutex this way. My understanding that if it is not contended, it does not require a syscall, so it should be fast. 

I'll add a note because as you said, I am not sure what exactly will be required in stall detection. I also don't know how to get rid of `Mutex` because `RefCell` does not seem to work in this case (where you need to return something under DashMap's mutex). `Arc<Stream>` may work though